pkg/cloudmeta/metadata.go

```diff
@@ -0,0 +1,83 @@
+// Copyright (C) 2024 ScyllaDB
+
+package cloudmeta
+
+import (
+	"context"
+	"errors"
+	"time"
+)
+
+// InstanceMetadata represents metadata returned by cloud provider.
+type InstanceMetadata struct {
+	InstanceType  string
+	CloudProvider CloudProvider
+}
+
+// CloudProvider is enum of supported cloud providers.
+type CloudProvider string
+
+// CloudProviderAWS represents aws provider.
+var CloudProviderAWS CloudProvider = "aws"
+
+// CloudMetadataProvider interface that each metadata provider should implement.
+type CloudMetadataProvider interface {
+	Metadata(ctx context.Context) (InstanceMetadata, error)
+}
+
+// CloudMeta is a wrapper around various cloud metadata providers.
+type CloudMeta struct {
+	providers []CloudMetadataProvider
+
+	providerTimeout time.Duration
+}
+
+// NewCloudMeta creates new CloudMeta provider.
+func NewCloudMeta() (*CloudMeta, error) {
+	const defaultTimeout = 5 * time.Second
+
+	awsMeta, err := NewAWSMetadata()
+	if err != nil {
+		return nil, err
+	}
+
+	return &CloudMeta{
+		providers: []CloudMetadataProvider{
+			awsMeta,
+		},
+		providerTimeout: defaultTimeout,
+	}, nil
+}
+
+// ErrNoProviders will be returned by CloudMeta service, when it hasn't been initialized with any metadata provider.
+var ErrNoProviders = errors.New("no metadata providers found")
+
+// GetInstanceMetadata tries to fetch instance metadata from AWS, GCP, Azure providers in order.
+func (cloud *CloudMeta) GetInstanceMetadata(ctx context.Context) (InstanceMetadata, error) {
+	if len(cloud.providers) == 0 {
+		return InstanceMetadata{}, ErrNoProviders
+	}
+	var mErr error
+	for _, provider := range cloud.providers {
+		meta, err := cloud.runWithTimeout(ctx, provider)
+		if err != nil {
+			mErr = errors.Join(mErr, err)
+			continue
+		}
+		return meta, nil
+	}
+
+	return InstanceMetadata{}, mErr
+}
+
+func (cloud *CloudMeta) runWithTimeout(ctx context.Context, provider CloudMetadataProvider) (InstanceMetadata, error) {
+	ctx, cancel := context.WithTimeout(ctx, cloud.providerTimeout)
+	defer cancel()
```

Comment:
cancel function is not needed here, you can skip it.
Context will be cancelled by the timeout.

Reply:
agree, is indeed not need here, but out linter doesn't like that idea `lostcancel: the cancel function returned by context.WithTimeout should be called, not discarded, to avoid a context leak (govet)` :) I also prefer not to ignore cancel functions, because if someone will extend this function, then it easy to miss that cancel is ignored.